torchsde/_brownian/brownian_path.py

```diff
@@ -58,119 +69,232 @@ def __init__(self,
+            self._ws.insert(idx, W_h1)
+            self._us.insert(idx, U_h1)
+        else:
+            h = self._ts[idx] - self._ts[idx - 1]
+            h1 = t - self._ts[idx - 1]
+            h2 = h - h1
+
+            W_h, U_h = self._ws[idx], self._us[idx]
+            W_h1, U_h1 = utils.brownian_bridge_augmented(self._w0, h1, h, W_h, U_h)
+
+            W_h2 = W_h - W_h1
+            U_h2 = U_h - U_h1 - h2 * W_h1
 
+            self._ws[idx] = W_h2
+            self._us[idx] = U_h2
+
+            self._ts.insert(idx, t)
+            self._ws.insert(idx, W_h1)
+            self._us.insert(idx, U_h1)
+        return idx
+
+    def _aggregate_W(self, idx_a: int, idx_b: int) -> torch.Tensor:
+        """Aggregate Brownian increments."""
+        if idx_b > idx_a:
+            return sum(self._ws[idx_a + 1: idx_b + 1])
+        return torch.zeros_like(self._w0)
+
+    def _aggregate_W_U(self, idx_a: int, idx_b: int) -> Tuple[torch.Tensor, ...]:
+        """Aggregate Brownian increments and space-time Levy area."""
+        W, U = self._ws[idx_a + 1], self._us[idx_a + 1]
+        for i in range(idx_a + 2, idx_b + 1):
+            U = U + self._us[i] + (self._ts[i] - self._ts[i - 1]) * W
+            W = W + self._ws[i]
+        return W, U
+
+    def _aggregate_W_U_A(self, idx_a: int, idx_b: int) -> Tuple[torch.Tensor, ...]:
+        """Aggregate Brownian increments, space-time Levy area, and space-space Levy area."""
+        W, U = self._aggregate_W_U(idx_a, idx_b)
+        h = self._ts[idx_b] - self._ts[idx_a]
+        H = utils.U_to_H(W, U, h)
+        A = utils.davie_foster_approximation(W, H, h, self.levy_area_approximation)
+        return W, U, A
+
+    def _point_eval(self, t: float) -> torch.Tensor:
+        idx_a, idx_b = 0, self._update_state(t)
+        return self._aggregate_W(idx_a, idx_b) + self._w0
+
+    def _interval_eval(self, ta: float, tb: float, return_U=False, return_A=False) -> TensorOrTensors:
+        if ta > tb:
+            raise RuntimeError(f"Query times ta={ta:.3f} and tb={tb:.3f} must respect ta <= tb.")
+
+        idx_a = self._update_state(ta)
+        idx_b = self._update_state(tb)
         if return_U:
             if return_A:
-                return W, U, A
+                return self._aggregate_W_U_A(idx_a, idx_b)
             else:
-                return W, U
+                return self._aggregate_W_U(idx_a, idx_b)
         else:
             if return_A:
+                W, U, A = self._aggregate_W_U_A(idx_a, idx_b)
                 return W, A
             else:
-                return W
+                return self._aggregate_W(idx_a, idx_b)
+
+    def __call__(self, ta, tb=None, return_U=False, return_A=False) -> TensorOrTensors:
+        ta = float(ta)
+        tb = float(tb) if tb is not None else tb
+        if ta < self._t0:
+            warnings.warn(f"Should have ta>=t0 but got ta={ta} and t0={self._t0}.")
+            ta = self._t0
+
+        if tb is None:
```

Comment:
Fixed for now. But I think eventually we should get rid of the if/else statements in `__call__`. These are annoying to read and just add overhead, since they are called all the time. Preferably, the arguments passed to `__init__` should specify everything that's needed.

Reply:
So the reason I arranged it like this was so:
- the same BM can be used with multiple schemes (as in the diagnostics)
- the default behaviour is just to return the increment.

I'm not sure what else would satisfy those two conditions.